File: src/main/java/slavsquatsuperstar/mayonez/DebugDraw.java
package slavsquatsuperstar.mayonez;

import slavsquatsuperstar.mayonez.physics2d.primitives.AlignedBoxCollider2D;
import slavsquatsuperstar.mayonez.physics2d.primitives.BoxCollider2D;
import slavsquatsuperstar.mayonez.physics2d.primitives.CircleCollider;
import slavsquatsuperstar.mayonez.physics2d.primitives.Collider2D;

import java.awt.*;
import java.util.ArrayList;
import java.util.List;

// TODO create IMGUI implementation
public class DebugDraw {

    private static final int STROKE_SIZE = 2;
    private static final Stroke stroke = new BasicStroke(STROKE_SIZE);
    private static final List<ShapeDrawer> shapes = new ArrayList<>(); // map to color?

    // Points

    public static void drawPoint(Vector2 position, Color color) {
        Logger.log("Point: %s", position);
        // Fill a circle with radius "STROKE_SIZE"
        Vector2 min = position;
        shapes.add(g2 -> {
            g2.setColor(color);
            g2.fillOval(toScreen(min.x) - STROKE_SIZE, toScreen(min.y) - STROKE_SIZE, STROKE_SIZE * 2, STROKE_SIZE * 2);
        });
    }

    public static void drawPointScreen(Vector2 position, Color color) {
        Logger.log("Point: %s", position);
        // Fill a circle with radius "STROKE_SIZE"
        Vector2 min = position.sub(new Vector2(STROKE_SIZE, STROKE_SIZE));
        shapes.add(g2 -> {
            g2.setColor(color);
            g2.fillOval((int) min.x, (int) min.y, STROKE_SIZE * 2, STROKE_SIZE * 2);
        });
    }

    // Lines

    public static void drawLine(Vector2 start, Vector2 end, Color color) {
        shapes.add(g2 -> {
            g2.setColor(color);
            g2.drawLine(toScreen(start.x), toScreen(start.y),
                    toScreen(end.x), toScreen(end.y));
        });
    }

    public static void drawLineScreen(Vector2 start, Vector2 end, Color color) {
        shapes.add(g2 -> {
            g2.setColor(color);
            g2.drawLine((int) start.x, (int) start.y, (int) end.x, (int) end.y);
        });
    }

    public static void drawVector(Vector2 direction, Vector2 origin, Color color) {
        drawLine(origin, origin.add(direction), color);
//        drawPoint(origin.add(direction), color);
    }

    // Shapes

    public static void drawShape(Collider2D shape, Color color) {
        if (shape instanceof CircleCollider)
            drawCircle((CircleCollider) shape, color);
        else if (shape instanceof AlignedBoxCollider2D)
            drawAABB((AlignedBoxCollider2D) shape, color);
        else if (shape instanceof BoxCollider2D)
            drawBox((BoxCollider2D) shape, color);
    }

    public static void drawShapeScreen(Collider2D shape, Color color) {
        if (shape instanceof CircleCollider)
            drawCircleScreen((CircleCollider) shape, color);
        else if (shape instanceof AlignedBoxCollider2D)
            drawAABBScreen((AlignedBoxCollider2D) shape, color);
        else if (shape instanceof BoxCollider2D)
            drawBoxScreen((BoxCollider2D) shape, color);
    }

    private static void drawCircle(CircleCollider circle, Color color) {
        shapes.add(g2 -> {
            g2.setColor(color);
            g2.drawOval(toScreen(circle.min().x), toScreen(circle.min().y),
                    toScreen(circle.radius * 2), toScreen(circle.radius * 2));
        });
    }

    private static void drawCircleScreen(CircleCollider circle, Color color) {
        shapes.add(g2 -> {
            g2.setColor(color);
            g2.drawOval((int) circle.min().x, (int) circle.min().y, (int) (circle.radius * 2), (int) (circle.radius * 2));
        });
    }

    private static void drawAABB(AlignedBoxCollider2D aabb, Color color) {
        shapes.add(g2 -> {
            g2.setColor(color);
            g2.drawRect(toScreen(aabb.min().x), toScreen(aabb.min().y), toScreen(aabb.width()), toScreen(aabb.height()));
        });
    }

    private static void drawAABBScreen(AlignedBoxCollider2D aabb, Color color) {
        shapes.add(g2 -> {
            g2.setColor(color);
            g2.drawRect((int) aabb.min().x, (int) aabb.min().y, (int) aabb.width(), (int) aabb.height());
        });
    }

    private static void drawBox(BoxCollider2D box, Color color) {
        Polygon obb = new Polygon();
        for (Vector2 point : box.getVertices())
            obb.addPoint(toScreen(point.x), toScreen(point.y));
        shapes.add(g2 -> {
            g2.setColor(color);
            g2.drawPolygon(obb);
        });
    }

    private static void drawBoxScreen(BoxCollider2D box, Color color) {
        Polygon obb = new Polygon();
        for (Vector2 point : box.getVertices()) obb.addPoint((int) point.x, (int) point.y);
        shapes.add(g2 -> {
            g2.setColor(color);
            g2.drawPolygon(obb);
        });
    }

    private static int toScreen(float world) {
        return Math.round(world * Game.currentScene().getCellSize());
    }

    // Helper Members

    public void render(Graphics2D g2) {
        if (!shapes.isEmpty()) {
            g2.setStroke(stroke);
            shapes.forEach(s -> s.draw(g2));
            shapes.clear();
        }
    }

    @FunctionalInterface
    private interface ShapeDrawer { // Maps a shape to a draw function
        void draw(Graphics2D g2);
    }

}

File: mayonez-base/src/main/java/slavsquatsuperstar/mayonez/graphics/DebugDraw.kt
package slavsquatsuperstar.mayonez.graphics

import slavsquatsuperstar.math.Vec2
import slavsquatsuperstar.mayonez.Mayonez
import slavsquatsuperstar.mayonez.annotations.EngineType
import slavsquatsuperstar.mayonez.annotations.UsesEngine
import slavsquatsuperstar.mayonez.event.Receivable
import slavsquatsuperstar.mayonez.physics.colliders.Collider
import slavsquatsuperstar.mayonez.physics.shapes.*
import slavsquatsuperstar.util.Colors
import java.awt.BasicStroke
import java.awt.Color
import java.awt.Graphics2D
import java.awt.Stroke
import java.awt.geom.AffineTransform
import java.awt.geom.Ellipse2D
import java.awt.geom.Line2D
import java.awt.geom.Rectangle2D
import kotlin.math.roundToInt

/**
 * Draws colliders and mathematical objects onto the screen. All methods use world coordinates, and shapes are centered
 * around the collider's position.
 *
 * @author SlavSquatSuperstar
 */
@UsesEngine(EngineType.AWT)
// TODO need z-indexing to work with renderer
object DebugDraw {

    private const val STROKE_SIZE = 2
    private val stroke: Stroke = BasicStroke(STROKE_SIZE.toFloat())
    private val shapes: MutableList<ShapeDrawer> = ArrayList() // map to Color

    // Game Loop Method
    fun render(g2: Graphics2D) {
        if (shapes.isNotEmpty()) {
            shapes.sortBy { s: ShapeDrawer -> s.priority.ordinal } // Sort shapes by priority, may case lag
            g2.stroke = stroke
            shapes.forEach { s: ShapeDrawer -> s.draw(g2) }
            shapes.clear()
        }
    }

    // Public Draw Methods
    /**
     * Draws a point onto the screen
     *
     * @param position where the point is located, in world coordinates
     * @param color    the color to use
     */
    @JvmStatic
    fun drawPoint(position: Vec2, color: Color?) {
        if (!Mayonez.started) return
        // Fill a circle with radius "STROKE_SIZE" in pixels
        val diameterPx = Vec2(STROKE_SIZE.toFloat() * 2f)
        shapes.add(ShapeDrawer(DrawPriority.POINT, ellipse(position.toScreen(), diameterPx), color, true))
    }

    /**
     * Draws a line segment onto the screen.
     *
     * @param start the segment's starting point, in world coordinates
     * @param end   the segment's ending point in, world coordinates
     * @param color the color to use
     */
    @JvmStatic
    fun drawLine(start: Vec2, end: Vec2, color: Color?) {
        if (!Mayonez.started) return
        val startPx = start.toScreen()
        val endPx = end.toScreen()
        shapes.add(ShapeDrawer(DrawPriority.LINE, Line2D.Float(startPx.x, startPx.y, endPx.x, endPx.y), color, false))
    }

    /**
     * Draws a vector onto the screen.
     *
     * @param origin    the vector's starting point, in world coordinates
     * @param direction how far away the vector's end point is, in world coordinates
     * @param color     the color to use
     */
    @JvmStatic
    fun drawVector(origin: Vec2, direction: Vec2, color: Color?) = drawLine(origin, origin.add(direction), color)

    // Draw Shapes
    /**
     * Draws the outline of a shape onto the screen.
     *
     * @param shape a [Collider]
     * @param color the color to use
     */
    @JvmStatic
    fun drawShape(shape: Shape?, color: Color?) {
        if (!Mayonez.started) return
        val drawColor = color ?: Colors.BLACK
        when (shape) {
            is Edge -> drawLine(shape.start, shape.end, drawColor)
            is Circle -> drawCircle(shape, drawColor, false)
            is Ellipse -> drawEllipse(shape, drawColor, false)
            is Rectangle -> drawRectangle(shape, drawColor, false)
            is Polygon -> drawPolygon(shape, drawColor, false)
        }
    }

    /**
     * Draws a shape onto the screen and fills in the interior.
     *
     * @param shape a [Collider]
     * @param color the color to use
     */
    @JvmStatic
    fun fillShape(shape: Shape?, color: Color?) {
        if (!Mayonez.started) return
        val drawColor = color ?: Colors.BLACK
        when (shape) {
            is Edge -> drawLine(shape.start, shape.end, drawColor)
            is Circle -> drawCircle(shape, drawColor, true)
            is Ellipse -> drawEllipse(shape, drawColor, true)
            is Rectangle -> drawRectangle(shape, drawColor, true)
            is Polygon -> drawPolygon(shape, drawColor, true)
        }
    }

    // Internal Draw methods

    private fun drawToScreen(g2: Graphics2D, color: Color?, shape: java.awt.Shape, fill: Boolean) {
        g2.color = color ?: Colors.BLACK
        if (fill) g2.fill(shape)
        else g2.draw(shape)
    }

    private fun drawCircle(circle: Circle, color: Color, fill: Boolean) {
        val diameterPx = Vec2(circle.radius * 2).toScreen()
        shapes.add(ShapeDrawer(DrawPriority.SHAPE, ellipse(circle.center().toScreen(), diameterPx), color, fill))
    }

    private fun drawEllipse(ellipse: Ellipse, color: Color, fill: Boolean) {
        if (ellipse.isCircle) return drawCircle(ellipse.boundingCircle(), color, fill)
        val centerPx = ellipse.center().toScreen()
        val shape = ellipse(centerPx, ellipse.size.toScreen())
        val rot = if (ellipse.isAxisAligned) shape else shape.rotate(centerPx, ellipse.angle)
        shapes.add(ShapeDrawer(DrawPriority.SHAPE, rot, color, fill))
    }

    private fun drawPolygon(polygon: Polygon, color: Color, fill: Boolean) {
        val shape = java.awt.Polygon()
        polygon.vertices.forEach { shape.addPoint(it.x.toScreen().roundToInt(), it.y.toScreen().roundToInt()) }
        shapes.add(ShapeDrawer(DrawPriority.SHAPE, shape, color, fill))
    }

    private fun drawRectangle(rect: Rectangle, color: Color, fill: Boolean) {
        val minPx = rect.min().toScreen()
        val centerPx = rect.center().toScreen()
        val shape = Rectangle2D.Float(minPx.x, minPx.y, rect.width.toScreen(), rect.height.toScreen())
        val rot = if (rect.isAxisAligned) shape else shape.rotate(centerPx, rect.angle)
        shapes.add(ShapeDrawer(DrawPriority.SHAPE, rot, color, fill))
    }

    // Helper Methods/Classes

    private fun ellipse(center: Vec2, size: Vec2): Ellipse2D {
        val min = center - size * 0.5f
        return Ellipse2D.Float(min.x, min.y, size.x, size.y)
    }

    private fun java.awt.Shape.rotate(center: Vec2, angle: Float): java.awt.Shape {
        val rotXf = AffineTransform.getRotateInstance(
            Math.toRadians(angle.toDouble()),
            center.x.toDouble(), center.y.toDouble()
        )
        return rotXf.createTransformedShape(this)
    }

    /**
     * Converts a coordinate from world to screen units.
     *
     * @return the corresponding screen pixel
     */
    private fun Float.toScreen(): Float = this * Mayonez.scene.scale

    /**
     * Converts a pair of coordinates from world to screen units.
     *
     * @return the corresponding screen pixels
     */
    private fun Vec2.toScreen(): Vec2 = this * Mayonez.scene.scale

    /**
     * The order an object should be drawn. Higher priority objects will be drawn later to be more visible.
     */
    private enum class DrawPriority {
        SHAPE, LINE, POINT
    }

    /**
     * Maps a shape to a draw function.
     */
    private class ShapeDrawer(
        val priority: DrawPriority,
        private val shape: java.awt.Shape,
        private val color: Color?,
        private val fill: Boolean
    ) : Receivable {
        fun draw(g2: Graphics2D) = onReceive(g2)

        override fun onReceive(vararg args: Any?) {
            val g2 = args[0]!! as Graphics2D
            g2.color = color ?: Colors.BLACK
            if (fill) g2.fill(shape)
            else g2.draw(shape)
        }
    }

}